Improved mapmri implementation with laplacian regularization and new …

rutgerfick opened the pull request:
…q-space scalar indices: q-space inverse variance and mean squared displacement. Also includes generalized cross-validation to find optimal regularization weights. Other convenience functions have been added.

arokem commented:
Nice one! @maurozucchelli : would this conflict in any way with the work you did on #733? I think we should probably merge that one first, before harmonizing with this. What do you both think?

rutgerfick commented:
I took #733 as the 'dipy' mapmri template where I added my own code for laplacian regularization and other scalar indices. So there is no need to harmonize that one first.

rutgerfick commented:
Ah sorry, I didn't see there was also some changes to a mapmri test in #733 . In any case I don't think the mapmri.py code here conflicts with #733.

arokem commented:
OK - let's still do this one-by-one, just to keep things nice and tidy. I suggest we still merge #733 first (once example is done there). Once that's done, you can rebase this one, to absorb the changes to the tests, and then maybe add to the example some of these maps as well. 

A few comments for now: 

1. Testing: you will need to write unit tests for all these new methods. Try following the examples in `test_mapmri.py`. In particular, I see that you have found a bug in the ordering of the tensor eigenvectors. In addition to fixing the bug (great catch!), it is necessary to add a test that would fail with this bug, so that it doesn't somehow creep back into our code-base (bugs have the tendency to do that). 

2. Do I understand correctly your comments that this implementation no longer requires cvx? That would be excellent news! If that is indeed the case, you should remove the import of cvx at the top of the file, and you can scrub away all the warnings about cvx, GPL license and all that (lines 225-231 in your implementation). Well done on that!

Garyfallidis commented:
Hi @rutgerfick very happy to see progress here. Can't wait to start playing with this. 
Before removing cvxopt we should see here some timings and some result comparisons. Is there any specific paper or application which shows that your approach is doing a better job?
Also @maurozucchelli I hope you are looking at this PR. We need your eyes on this too.

arokem commented:
I'd say that if we don't see significant performance regression (I don't expect that, it's not like cvx is that fast...), and it passes this test: https://github.com/nipy/dipy/blob/master/dipy/reconst/tests/test_mapmri.py#L49, then we should be fine. What do you think @Garyfallidis?

Garyfallidis commented:
Completely agree and maybe the same idea can be used for SHORE.

arokem commented:
That would be just too beautiful.

rutgerfick commented:
Sorry maybe I was unclear. 

The positivity constraint is still using cvxopt, I didnt' change anything with that.
The added value of this implementation is that you can use the laplacian regularization on its own and not use the positivity constraint. In a work that's under review now I show that using the Laplacian you get more accurate results than the positivity constraint in terms of ODF reconstruction. Also, it's much faster :)

arokem commented:
Hrrm - IANAL, but I think that we would still need to leave the GPL warning
intact then, to be on the safe side.

On Mon, Oct 19, 2015 at 10:45 AM, rutgerfick <notifications@github.com>
wrote:

> Sorry maybe I was unclear.
>
> The positivity constraint is still using cvxopt, I didnt' change anything
> with that.
> The added value of this implementation is that you can use the laplacian
> regularization on its own and not use the positivity constraint. In a work
> that's under review now I show that using the Laplacian you get more
> accurate results than the positivity constraint in terms of ODF
> reconstruction. Also, it's much faster :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-149294266>.
>

maurozucchelli commented:
I think the order of the merge is not important. My last pull request just
added some scalar maps, while Rutger's improved the fitting.

On Mon, Oct 19, 2015 at 7:49 PM, Ariel Rokem <notifications@github.com>
wrote:

> Hrrm - IANAL, but I think that we would still need to leave the GPL warning
> intact then, to be on the safe side.
>
> On Mon, Oct 19, 2015 at 10:45 AM, rutgerfick <notifications@github.com>
> wrote:
>
> > Sorry maybe I was unclear.
> >
> > The positivity constraint is still using cvxopt, I didnt' change anything
> > with that.
> > The added value of this implementation is that you can use the laplacian
> > regularization on its own and not use the positivity constraint. In a
> work
> > that's under review now I show that using the Laplacian you get more
> > accurate results than the positivity constraint in terms of ODF
> > reconstruction. Also, it's much faster :)
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/nipy/dipy/pull/740#issuecomment-149294266>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-149295572>.
>

samuelstjean commented:
Do you still get strictly positive estimates without the constraint when you only use regularisation or it's suggested to use both at the same time?

rutgerfick commented:
The Laplacian only induces a smoothness into the fitting of the signal, so it does not strictly impose positivity in the EAP in any way. However, this smoothness does reduce the spurious behavior in the EAP (similarly but better than the low-pass filters in dipy's 3D-SHORE code). Using them both would make sure you positively recover all metrics, but I would like to give you these sidenotes on the Laplacian on HCP data and a dMRI phantom:

In terms of scalar indices, it typically recovers smooth, positive maps in the white and grey matter, only finding negative values near the skull.
In terms of ODFs, the laplacian makes the peaks a little bit less sharp, but reduces the underestimation of the crossing angle that is inherent in MAP-MRI.
In terms of signal fitting, it performs equally or better than the positivity constraint, especially in the extrapolation.
In terms of speed, it's a lot faster than the positivity constraint, especially for higher radial orders.

I'll put in a citation to these results as soon as my paper is out, but this was what I found. Still, you can use both and compare results yourself ;)

arokem commented:
Hey @rutgerfick - please rebase this on top of master (let me know if you need any help doing that). Once you have added tests for the new functions, I'd be happy to take another look.

rutgerfick commented:
Of course, I might have time to figure out how to do do this and make some
tests tomorrow :)

2015-10-22 16:26 GMT+02:00 Ariel Rokem <notifications@github.com>:

> Hey @rutgerfick <https://github.com/rutgerfick> - please rebase this on
> top of master (let me know if you need any help doing that). Once you have
> added tests for the new functions, I'd be happy to take another look.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-150240523>.
>

Garyfallidis commented:
ping

rutgerfick commented:
Hello @arokem @Garyfallidis. I have been working hard and added tests for all estimated q-space indices (RTOP, RTAP, RTPP, MSD, and QIV) and Laplacian regularization.

However, I have also added some new functionality, including their tests. If you prefer, I can remove these so that this pull-request is more clean, but I would like to leave this decision to you.

More specifically, the added functionality includes:
- Estimation of Non-Gaussianity (NG), perpendicular NG and parallel NG, with corresponding tests.
- The spherical MAP-MRI implementation, which is automatically used when anisotropic_scaling is set to False, greatly boosting fitting speed.
- Test showing equality of the fitted signal between Cartesian MAP-MRI using isotropic scaling and the spherical implementation (i.e., showing equality with 3D-SHORE).
- The Laplacian regularization for the spherical implementation and its test.
- Estimation of all q-space indices in in the spherical implementation, with the corresponding tests.
- Function odf_sh that analytically produces the spherical harmonic coefficients of the ODF instead of the discrete sphere function for the isotropic implementation.

Please let me know your preference :)
Rutger

arokem commented:
Hey @rutgerfick - thanks for all your work on this! 

Did you commit and push the tests? I can't see them here. 

My tendency is to say that we plow ahead with the full PR, including all the additional functionality you mentioned. But bear with us, as this might take a little while to review. In particular,  we might not have time to give this a proper review before upcoming release (Friday!). 

For the time being, could you please rebase this on top of the current master branch? There seem to be some conflicts.

rutgerfick commented:
No problem! I didn't push it yet because I didn't know if you wanted all the extra stuff.

Could you please help me with how to rebase to the current master branch? 
I suppose I'll then deal with the conflicts and push it for you guys :)

arokem commented:
Here's a good tutorial on rebasing: https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

rutgerfick commented:
Hi @arokem, 
I rebased on the dipy_master branch and added even more tests to make sure everything is working as it should. 

Can you tell me why it still says that this branch has conflicts if I have already rebased? What else should I do?

arokem commented:
I don't know. Hard to tell from here... Just checking: did you push your rebased branch?

rutgerfick commented:
When i push and pull it says I'm up-to-date either way. Also in the large list of commits above my last post it shows my commits twice now. Maybe because I tried to rebase twice?

arokem commented:
Yeah. There does seem to be something wrong with the rebase -- the history here includes commits that are already in master. Might require a more delicate approach... an interactive rebase, perhaps.

Might be more amenable to handling this face-to-face (over skype). Could you get in touch (gh-username@gmail.com)? Let's try to find a time that would work for both of us to do that.

arokem commented:
Hey @rutgerfick - thanks for taking the time to rebase this. Please take a look at the Travis tests. It seems that the tests are not yet running smoothly: https://travis-ci.org/nipy/dipy/jobs/109931283. Hopefully not something we broke while rebasing!

rutgerfick commented:
Yeah we broke something. I'll fix it. hold on.

rutgerfick commented:
Any idea why it says that checks have failed? Locally all the mapmri tests work fine. In the details it doesn't seem that any of the fails have anything to do with the mapmri code...

arokem commented:
You can see the test runs here:https://travis-ci.org/nipy/dipy/builds/110076846

I'll also take a look in a little bit (run it locally on my machine and try to debug).

arokem commented:
Yeah - I am mystified. Tests run just fine on my machine.

arokem commented:
Gotcha. `test_matpmri_odf` made its way back in?

rutgerfick commented:
For some reason the other tests gave errors in the example in the documentation lines of mapmri.py. Now i copied the example of shore.py into mapmri.py and just switched the model to mapmri. Hopefully it will work now.

arokem commented:
Yes. We want the examples in the docstrings to work as well! So the travis machines are set to test with

    nosetests --with-doctest

rutgerfick commented:
Alright travis now passes all tests, except for the one that requires cvxopt to do the positivity constraint, which is supposed to happen. Now it's all good for now right?

arokem commented:
We would need to "protect" this test from being executed when cvxopt is not present. For an example of how that's done, look to this line: 

https://github.com/nipy/dipy/blob/master/dipy/reconst/tests/test_sfm.py#L22

In this case, the model requires `sklearn` and the test is skipped if the machine does not have `sklearn` installed.

Garyfallidis commented:
Hi @rutgerfick and thank you for this work and also apologies for delaying to review this and give you feedback. I tried to run the existing example of mapmri and it fails with an unexpected keyword argument. Here is the output

```
/home/eleftherios/Devel/dipy/doc/examples/reconst_mapmri.py in <module>()
     55 radial_order = 4
     56 map_model = MapmriModel(gtab, radial_order=radial_order,
---> 57                         lambd=2e-1, eap_cons=False)
     58 
     59 """

TypeError: __init__() got an unexpected keyword argument 'eap_cons'
> /home/eleftherios/Devel/dipy/doc/examples/reconst_mapmri.py(57)<module>()
     56 map_model = MapmriModel(gtab, radial_order=radial_order,
---> 57                         lambd=2e-1, eap_cons=False)
     58 
```

Garyfallidis commented on dipy/reconst/tests/test_mapmri.py:
```diff
@@ -60,45 +188,463 @@ def test_mapmri_metrics():
+    assert_equal(laplacian_norm_laplacian < laplacian_norm_unreg, True)
+
+
+def test_shore_odf(radial_order=6):
```
Shouldn't this be ``test_mapmri_odf``?

rutgerfick commented:
Ah i didn't touch the example at all. I'll adapt it tomorrow :)
On Mar 11, 2016 11:37 PM, "Eleftherios Garyfallidis" <
notifications@github.com> wrote:

> Hi @rutgerfick <https://github.com/rutgerfick> and thank you for this
> work and also apologies for delaying to review this and give you feedback.
> I tried to run the existing example of mapmri and it fails with an
> unexpected keyword argument. Here is the output
>
> /home/eleftherios/Devel/dipy/doc/examples/reconst_mapmri.py in <module>()
>      55 radial_order = 4
>      56 map_model = MapmriModel(gtab, radial_order=radial_order,
> ---> 57                         lambd=2e-1, eap_cons=False)
>      58
>      59 """
>
> TypeError: __init__() got an unexpected keyword argument 'eap_cons'
> > /home/eleftherios/Devel/dipy/doc/examples/reconst_mapmri.py(57)<module>()
>      56 map_model = MapmriModel(gtab, radial_order=radial_order,
> ---> 57                         lambd=2e-1, eap_cons=False)
>      58
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-195586062>.
>

Garyfallidis commented:
Cool. Check also your coverage. It's already quite high but you maybe missing some cornercases. Here is the summary of missing lines from the travis bot https://travis-ci.org/nipy/dipy/jobs/112036719

```
dipy/reconst/mapmri.py                               707     43    248     25    92%   204-205, 211, 217, 228, 255-266, 283-286, 302, 334, 443-445, 694-696, 722-724, 762-764, 789, 797-798, 824-833, 1577, 1580, 1876, 203->204, 206->212, 210->211, 216->217, 225->228, 251->255, 282->283, 301->302, 305->313, 333->334, 416->421, 442->443, 448->452, 689->693, 693->694, 717->721, 721->722, 757->761, 761->762, 786->789, 796->797, 823->824, 1576->1577, 1579->1580, 1875->1876
```

Garyfallidis commented on dipy/reconst/tests/test_mapmri.py:
```diff
@@ -52,7 +163,24 @@ def test_mapmri_metrics():
+    mapfit = mapm.fit(S)
+    c_map = mapfit.mapmri_coeff
+
+    R = mapfit.mapmri_R
```
R and mu are assigned but not used.

Garyfallidis commented:
I cannot see the diff of the file dipy/reconst/mapmri.py because it's too large and github avoids showing it. So, I will have to give you non-interactive feedback. Will do that, no worries.

Garyfallidis commented:
Line 76   ``DTI_scale_estimation=True``, should be ``dti_scale_estimation=True`` we try to always use small letters in parameters.

Garyfallidis commented:
Can you summarize what is the added value of using laplacian regularization? What this improves for MAPMRI? 
Also what is the added value of using GCV (generalized cross validation)?
To reduce the angular bias should we use laplacian regularization or this is only related only to anisotropic scaling? I mean that the laplacian regularization should increase angular accuracy. Correct?

Please clarify lines 143-146.

Garyfallidis commented:
Line 264 Warning: line break before binary operator

Garyfallidis commented:
Line 674 trailing whitespace.

Garyfallidis commented:
Lines 815-816 align docstring to the left.

Garyfallidis commented:
If you think that the GCV function (line 1846) is generic enough we should consider having it out of mapmri.py There are also some other cross validation functions in dipy/reconst/cross_validation.py

Garyfallidis commented:
After looking and correcting the mapmri example you need to make sure also that the example is connected to the documentation generation.
See files ``dipy/doc/examples_index.rst`` and ``doc/examples/valid_examples.txt``.

Garyfallidis commented:
Finally, I want to you to think of what would be the best usage example for MAPMRI. And if the current example does not cover that then we need to perhaps add another example which would promote the method better. So, for you what example would be the ideal promotion for MAPMRI and its many maps? Human brain, animal data of high resolution or something else? Thanks again for the great work.

rutgerfick commented:
Thanks for all the feedback @Garyfallidis ! Let me prepare a detailed response to all your comments after the MICCAI deadline next Thursday :)

arokem commented on dipy/reconst/tests/test_mapmri.py:
```diff
@@ -1,50 +1,161 @@
+    q_par = qvec[:, 0]
+    E = callaghan_perpendicular(q_perp, R) * gaussian_parallel(q_par, tau)
+    E[q_norm == 0.] = 1.
+    return E
```
These functions could be really useful elsewhere. Maybe put them in the `sims` module, instead of being "buried" in here?

arokem commented on dipy/reconst/tests/test_mapmri.py:
```diff
@@ -1,50 +1,161 @@
+    # test the signal reconstruction
+    S = S / S[0]
+    nmse_signal = np.sqrt(np.sum((S - S_reconst) ** 2)) / (S.sum())
+    assert_almost_equal(nmse_signal, 0.0, 3)
```
Great!

arokem commented on dipy/reconst/tests/test_mapmri.py:
```diff
@@ -60,45 +188,463 @@ def test_mapmri_metrics():
+    l1, l2, l3 = [0.0015, 0.0003, 0.0003]
+    S, _ = generate_signal_crossing(gtab, l1, l2, l3)
+
+    radius_max = 0.02  # 40 microns
```
The comment doesn't agree with the code, I believe.

arokem commented on dipy/reconst/tests/test_mapmri.py:
```diff
@@ -60,45 +188,463 @@ def test_mapmri_metrics():
+def test_mapmri_metrics_anisotropic(radial_order=6):
+    gtab = get_gtab_taiwan_dsi()
+    l1, l2, l3 = [0.0015, 0.0003, 0.0003]
+    S, _ = generate_signal_crossing(gtab, l1, l2, l3, angle2=0)
```
When angle2=0, this is no longer a crossing? Do I understand correctly? Might be less confusing here to use the single tensor functions from dipy.sims.voxel

rutgerfick replied:
This has been changed

arokem commented on dipy/reconst/tests/test_mapmri.py:
```diff
@@ -60,45 +188,463 @@ def test_mapmri_metrics():
+
+    # test MAPMRI q-space indices
+
+    mapm = MapmriModel(gtab, radial_order=radial_order,
```
General question: would it make sense to test different radial orders? I see that you set this in the function signature in many places as a default to 6, but I wonder whether it would make sense to iterate (at least in one of the tests) over several plausible values.

rutgerfick replied:
I added a test over radial order. You can see that the fitting error decreases as radial order increases.

Garyfallidis commented:
Hi @rutgerfick we can't wait to have this merged! Are you on it?

Garyfallidis commented:
It will be exciting to have the entire MAPMRI experience :)

arokem commented:
You should probably also update the reference to your newly published
article (congratulations!)

On Sat, Apr 9, 2016 at 4:09 AM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> It will be exciting to have the entire MAPMRI experience :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-207771205>
>

rutgerfick commented:
Hi guys!
I'm on it. Sorry to be absent for a while, I took a little break after the MICCAI deadline.
I'll have the final comments implemented shortly :)

Garyfallidis commented:
Good to hear that you are back on this. Congrats for the paper too. :+1:

Garyfallidis commented:
Hi @rutgerfick I met with Rachid a couple of days ago and I was very glad to see that DIPY is steadily growing at INRIA and that he is very positive of continuing our collaborations in the same codebase.  

Now, you still have some conflicts with this PR. We need to have this PR ready relatively soon so that we can promote it during OHBM where we are going to have an exhibition stand and a workshop. Let us know if we can help somehow. Also, I will need to ask you to arrange a developers meeting (g+ hangout) where you can give us a bit more information about the updates given in this PR and the actual potential of your method. 

Let me know what you think.

rutgerfick commented:
Hi @Garyfallidis  @arokem,

I included all of your comments and added an extensive examples section that illustrates the added value of the laplacian and how it relates to estimating all scalar indices and the ODF. 

I want to resolve the conflicts, but if i do
git merge dipy_master
it tells me I'm already up-to-date. what should I do?

rutgerfick commented:
Regarding the developers meeting, I'm in France, what are your time zones? :)

rutgerfick commented:
One question on the examples. For the values of the q-space indices to make any sense I need to input the big_delta and small_delta in the gradient_table(). However, I can't seem to find these values of the cenir data that was used in the example previously. 
Do you know what their values are or where to find it?

arokem commented:
The full description of the data set (including these details) is here:
https://digital.lib.washington.edu/researchworks/handle/1773/33311?show=full

On Wed, Apr 27, 2016 at 6:25 AM, rutgerfick <notifications@github.com>
wrote:

> One question on the examples. For the values of the q-space indices to
> make any sense I need to input the big_delta and small_delta in the
> gradient_table(). However, I can't seem to find these values of the cenir
> data that was used in the example previously.
> Do you know what their values are or where to find it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-215082035>
>

rutgerfick commented:
Hi @Garyfallidis, I was thinking about what you said about an optimal example for mapmri's maps.
The current example uses a good multi-shell data set with many directions, but since the resolution isn't too good the maps don't really look as impressive as they could be.

Would it be possible to have some high-resolution multi-shell data set as an example? I think that would really boost the impact the images would make.

Garyfallidis commented:
Sure, tell us what you want. Do you have an example dataset in mind? Voxel size, number of directions, human or animal? Is there a public dataset that we could use or we need to acquire some datasets and share them?

rutgerfick commented:
The only public multi-shell data set with 1.25 mm^3 resolution is the WU-MINN HCP data. Do you know of anything higher resolution than that?

rutgerfick commented:
Also, I finished some code that is able to save and load fitted mapmri object with all of the extra data (scale functions, rotations etc).

Imagine that you fitted some big data set (some hours of computation), got the odfs or something, and then later decide you also want to have the RTOP or the odfs with a different sharpening factor. You'd have to fit the whole thing again and wait. Using this function you can just save the fitted object and load it, and use it exactly as if you just fitted. Do you want me to add it here or later?

Garyfallidis commented:
This PR is already quite large. Let's have this first version first and then for additions and better datasets we do that in the next PR. Make sure you resolve some of the current conflicts and get this PR ready to be merged. 

Also about the dataset. I don't know of a public dataset at a higher resolution that the HCP but maybe others here know. I will ask around. But this doesn't need to happen in this PR. But we should definitely improve the tutorials with better datasets :)

Garyfallidis commented:
Hi @rutgerfick are you ready for the review?

rutgerfick commented:
Hi @Garyfallidis,
Sure, what does the review entail?

Garyfallidis commented:
For now nothing more than looking again at your code and working towards giving you final comments and corrections. Then we may ask you for a hangout at some point if some things are not clear. Will get back to you asap. Thx!

Garyfallidis commented on doc/examples/reconst_mapmri.py:
```diff
@@ -31,36 +46,316 @@
+"""
+radial_order = 8
+map_model_both_iso = mapmri.MapmriModel(gtab, radial_order=radial_order,
+                          laplacian_regularization=True,
```
pep 8 parameters not properly aligned

rutgerfick commented:
Alright, everything should be good now. The example can be improved later, as you said.

arokem commented on dipy/reconst/mapmri.py:
```diff
@@ -261,97 +462,490 @@ def odf(self, sphere, s=0):
+            E = self.predict(gtab)
+        return E
+
+    def predict(self, qvals_or_gtab, S0=1.):
```
We're trying to be more uniform about this, setting S0 per default to 100 (https://github.com/nipy/dipy/pull/1063)

arokem commented:
It seems that there are still some issue on older versions of scipy (https://travis-ci.org/nipy/dipy/jobs/136408673#L2267-L2282) with you call to `fmin_l_fbgs_b`. Maybe you can try using the core.optimize.Optimizer object (https://github.com/nipy/dipy/blob/master/dipy/core/optimize.py#L22)? It was written with support for older versions of scipy in mind.

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
-data contains the voxel data and gtab contains a GradientTable
-object (gradient information e.g. b-values). For example, to show the b-values
-it is possible to write print(gtab.bvals).
+The MAPMRI Model can now be instantiated. The radial_ordermdetermines the
```
"The radial_orderdetermines..."=> "The ``radial_order``  parameter determines..."

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
-object (gradient information e.g. b-values). For example, to show the b-values
-it is possible to write print(gtab.bvals).
+The MAPMRI Model can now be instantiated. The radial_ordermdetermines the
+expansion order of the basis, i.e., how manymbasis functions are used to
```
"manymbasis" => "many basis"

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+purposes, as the anisotropic basis has a bias towards smaller crossing angles
+in the ODF. For signal fitting and estimation of scalar quantities the
+anisotropic basis is suggested. The choice can be made by setting
+anisotropic_scaling=True or False.
```
Add double quotes around code: "``anisotropic_scaling=True`` or ``anisotropic_scaling=False``"

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+anisotropic basis is suggested. The choice can be made by setting
+anisotropic_scaling=True or False.
+
+First, we must select the method of regularizing and/or constraining the basis
```
"... method of regularization..."

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+
+First, we must select the method of regularizing and/or constraining the basis
+fitting.
+- laplacian_regularization=True makes it use laplacian regularization
```
"``laplacian_regularization=True``"

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+fitting.
+- laplacian_regularization=True makes it use laplacian regularization
+  [Fick2016]_. This method essentially reduces spurious oscillations in the
+  reconstruction by minimizing the laplacian of the fitted signal.
```
Should "Laplacian" be written with a capital letter at the beginning (considering it's someone's name)?

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+  [Fick2016]_. This method essentially reduces spurious oscillations in the
+  reconstruction by minimizing the laplacian of the fitted signal.
+  Several options can be given to select the regularization weight:
+      -regularization_weighting="GCV" uses generalized cross-validation
```
=> "``regularization_weighting="GCV"``"

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+  Several options can be given to select the regularization weight:
+      -regularization_weighting="GCV" uses generalized cross-validation
+       [Craven1978]_ to find an optimal weight from a standard range.
+      -regularization_weighting=0.2 for example would ommit the GCV and
```
Again: code should be surround by double backticks.

arokem replied:
"ommit" => "omit"

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+      -regularization_weighting="GCV" uses generalized cross-validation
+       [Craven1978]_ to find an optimal weight from a standard range.
+      -regularization_weighting=0.2 for example would ommit the GCV and
+       just set it to 0.2 (found to be reasonable in HCP data [Fick2016]).
```
"[Fick2016]" => "[Fick2016]_"

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+       [Craven1978]_ to find an optimal weight from a standard range.
+      -regularization_weighting=0.2 for example would ommit the GCV and
+       just set it to 0.2 (found to be reasonable in HCP data [Fick2016]).
+      -regularization_weighting=np.array(weights) would make the GCV use
```
Surround code by double backticks

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+       just set it to 0.2 (found to be reasonable in HCP data [Fick2016]).
+      -regularization_weighting=np.array(weights) would make the GCV use
+       a custom range to find an optimal weight.
+- positivity_constraint=True makes it use the positivity constraint on the
```
Surround code by double backticks

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+  diffusion propagator [Ozarslan2013]_. This method constrains the final
+  solution of the diffusion propagator to be positive at a set of discrete
+  points, since negative values should not exist.
+      -the pos_grid and pos_radius affect the location and number of
```
pos_grid and pos_radius should be surrounded by double backticks

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+- Q-space Inverse Variance (QIV) is a measure of variance in the signal, which
+  is said to have higher contrast to white matter than the MSD
+  [Hosseinbor2013]_. We also showed that QIV has high sensitivity to tissue
+  composition change in a simulation study [Fick2016b].
```
=> "[Fick2016b]_" (with the underscore!)

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+  the probability that a proton will be along the axis of the main eigenvector
+  of a diffusion tensor during both diffusion gradient pulses. RTAP has been
+  related to the apparent axon diameter [Ozarslan2013, Fick2016]_ under several
+  strong hypothesis on the tissue composition and acquisition protocol.
```
"hypothesis" => "assumptions"?

arokem commented on doc/examples/reconst_mapmri.py:
```diff
@@ -32,35 +47,315 @@
+perpendicular NG and parallel NG. The NG ranges from 1 (completely
+non-Gaussian) to 0 (completely Gaussian). The overall NG of a voxel is always
+higher or equal than each of its components. It can be seen that NG has low
+values in the CSF and higher in the white matter.
```
Would be interesting to compare NG and MK. Should be really easy to do once this is merged!

cc: @RafaelNH

arokem commented on dipy/reconst/mapmri.py:
```diff
@@ -135,74 +256,154 @@ def __init__(self,
+                        self.radial_order, mu[0], constraint_grid)
+                    K = K_dependent * self.pos_K_independent
+
+            data = np.asarray(data / data[self.gtab.b0s_mask].mean())
```
Is this where you are hitting #1074?

arokem replied:
Should we add a `np.isscalar` on the output here @matthew-brett ? It would ber really ugly to special-case this to what is essentially just a bug in one version of Numpy.

rutgerfick replied:
Yeah this is exactly that. This is very ugly so any suggestion to deal with it more generally would be great.

arokem replied:
Would be good to add a test with a memmap here (and we might want to add this in other places as well...), so that we can be sure that we fix this.

samuelstjean commented:
Could be useful to have that directly in the reconst multivoxel fit part,
if it does not break anything. I currently set all my nibabel topmost
import to convert as an array upon loading since this issue was giving me
problems in other places also, but fixing it there would make it work for
any reconst object afterward.

2016-06-13 13:08 GMT+02:00 Coveralls <notifications@github.com>:

> [image: Coverage Status] <https://coveralls.io/builds/6570660>
>
> Changes Unknown when pulling *b1e00c4
> <https://github.com/nipy/dipy/commit/b1e00c4dc0d295c7baaf4cc922e3b957a2c74385>
> on AthenaEPI:athena_mapmri* into ** on nipy:master**.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-225551961>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AC4-6FChKBsAKnHi9HLeEWj8T0fv9JMXks5qLTougaJpZM4GRWmI>
> .
>

arokem commented:
But we'd need to check whether you are on numpy 1.11 or not, no? Which would be a pain in the neck. Maybe we can just ask people using numpy 1.11 not to do that? :-)

samuelstjean commented:
Nope, the exact line I use is 
data = np.asarray(vol.get_data(caching='unchanged'))  # To force ndarray instead of memmaps

Or as here just data = np.asarray(data) in this file https://github.com/nipy/dipy/blob/master/dipy/reconst/base.py#L37

But then I thought doing that might lead to people complaining that their hcp processing now takes 4go of ram for their hcp dataset (remember the dti takes too much memory issue?) and is plain impossible to run, so not a good idea in perspective.

rutgerfick commented:
So, any more corrections? :)

samuelstjean commented:
Well, I have some data which gives me degenerated metrics near the corpus
callosum and negative values near the ventricules, but I did not find out
yet if it's just bad data or maybe the fact that I did not enforce
positivity causing the problem.

2016-06-14 10:36 GMT+02:00 rutgerfick <notifications@github.com>:

> So, any more corrections? :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-225815380>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AC4-6A-XHt9FBGKtkM_tlt8Hj85I-3YVks5qLmf5gaJpZM4GRWmI>
> .
>

rutgerfick commented:
Are degenerate and negative the same?

If you use the "GCV" option it'll find the optimal regularization weight for that particular voxel. However, this optimal weight can be very low in voxels that are very anisotropic or decay quickly (e.g. corpus callosum or ventricles). If you set the regularization weight a bit higher it will in almost all cases become positive.

However, the Laplacian does not guarantee positivity! I also note this in the example. When you care about q-space indices, the best results are obtained by using a combination of both laplacian with either GCV or a lower weight and the positivity constraint, where in this case you can use fewer constraint points since the solution is smoother with the laplacian.

However, if you only care about signal interpolation for example, the laplacian GCV does a better job than the positivity constraint, as I recently found out.

samuelstjean commented:
Actually a bit of both, near the CC I get values which are much higher than
the rest (so high they screw up normalisation for visualisation, so they
are easy to spot) and negative metrics values (rtap/rtop) near the
ventricules.

I knew about the fact it does not enforce positivity (it was mentioned long
ago), but they are also not mutually exclusive. I tried once to have both,
but it was either taking ridiculously long (> 1 day compared to around 2-3
hours for laplacian regul) or it most probably hanged in some computation
with openmp/multiprocessing (which is a linalg issue, so not super
relevant). I guess I could give it a retry later this weekend to see, but
it should probably fix it.

I think I was on the default of 0.2 everywhere for this one (using the code
you sent me, so no option to save the object, but according to the script
should be that).

2016-06-14 10:59 GMT+02:00 rutgerfick <notifications@github.com>:

> Are degenerate and negative the same?
>
> If you use the "GCV" option it'll find the optimal regularization weight
> for that particular voxel. However, this optimal weight can be very low in
> voxels that are very anisotropic or decay quickly (e.g. corpus callosum or
> ventricles). If you set the regularization weight a bit higher it will in
> almost all cases become positive.
>
> However, the Laplacian does not guarantee positivity! I also note this in
> the example. When you care about q-space indices, the best results are
> obtained by using a combination of both laplacian with either GCV or a
> lower weight and the positivity constraint, where in this case you can use
> fewer constraint points since the solution is smoother with the laplacian.
>
> However, if you only care about signal interpolation for example, the
> laplacian GCV does a better job than the positivity constraint, as I
> recently found out.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-225821022>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AC4-6IsEpIB6qa5d-zoh9vy-sKwI3JGKks5qLm2MgaJpZM4GRWmI>
> .
>

rutgerfick commented:
once you guys decide to accept this pull request I'll do a second one right away that allows you to save fitted objects. It takes away a lot of the pressure to be 'faster' since you can fit and save it for later use.

Anyway, if you need me to change anything let me know. Also, the 'coveralls' thing tells me i'm 84% covered. Should that be 100%? how to get it there?

arokem commented:
I believe the 84% from coveralls is for the entire package. Coverage of the mapmri module is at 87% (https://travis-ci.org/nipy/dipy/jobs/137204729#L2386). That's actually not a great number. If you want to know which statements are not currently covered by tests, you should run `nosetests --with-coverage --cover-package=dipy` (see also: http://nose.readthedocs.io/en/latest/plugins/cover.html).

@matthew-brett, @Garyfallidis : do you have any thoughts on the array/memmap situation? The current solution seems like overkill to me. I would like for this to be taken as a final recourse only for numpy 1.11. What do you think?

Garyfallidis commented:
Hi @arokem and @rutgerfick. The problem of saving the fit results has been discussed many times in the past and when it will be implement it should support as many reconstruction modules as possible. Hopefully all of them if possible. There is no point at this stage allowing this only for MAPMRI. We need to think the general design. Of course we need to save the fitted parameters. @MrBago had made an interesting idea of how to parallelize the fit too. Memory, parallelization and speed are all components that need to be considered carefully. As is saving temporary the fit parameters to be used again at a later stage.

@arokem what do you mean with "do you have any thoughts on the array/memmap situation? The current solution seems like overkill to me". Which solution? Please be more specific. Apologies, that I don't remember the exact issue.

Garyfallidis commented:
@rutgerfick these are you missing lines. The coverage is 89% on my machine. Please make sure that something is not critical and try to go beyond 90%. Also @arokem I noticed that the SFM tests have dropped for some reason.

```
Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
dipy.reconst.mapmri     758     84    89%   210-211, 222, 228, 234, 239-248, 256, 316, 355, 365-367, 370, 392-394, 437, 443, 501-503, 753-755, 781-783, 821-823, 854-863, 873, 881-882, 908-917, 1459-1479, 1485-1503, 1661, 1664, 1950
```

@rutgerfick your example has rendering issues. You need to add a line space between all code snippets and all comments in the example. See line 351 for example.

Garyfallidis commented:
Also cvxopt should be optional and an alternative should be used instead. But maybe there is no alternative. If there is no alternative then the tests need to be skipped when cvxopt is not available. Let us know if you need some help with that. Keep it up @rutgerfick! We are getting there!

arokem commented:
On Tue, Jun 14, 2016 at 8:06 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> Hi @arokem <https://github.com/arokem> and @rutgerfick
> <https://github.com/rutgerfick>. The problem of saving the fit results
> has been discussed many times in the past and when it will be implement it
> should support as many reconstruction modules as possible. Hopefully all of
> them if possible. There is no point at this stage allowing this only for
> MAPMRI. We need to think the general design. Of course we need to save the
> fitted parameters. @MrBago <https://github.com/MrBago> had made an
> interesting idea of how to parallelize the fit too. Memory, parallelization
> and speed are all components that need to be considered carefully. As is
> saving temporary the fit parameters to be used again at a later stage.
>
> <https://github.com/arokem>
>
This is not in this PR. Let's discuss it when it becomes relevant.


> @arokem <https://github.com/arokem> what do you mean with "do you have
> any thoughts on the array/memmap situation? The current solution seems like
> overkill to me". Which solution? Please be more specific. Apologies, that I
> don't remember the exact issue.
>
>
https://github.com/nipy/dipy/issues/1074

And the solution in this commit:

https://github.com/nipy/dipy/pull/740/commits/b1e00c4dc0d295c7baaf4cc922e3b957a2c74385

This would solve the problem for numpy 1.11, but might be a burden in any
other version of numpy (unless I am misunderstanding the situation).




> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-226076733>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAHPNlj2FGIg0T6HZMwaRleg_OPYnWwwks5qL2w-gaJpZM4GRWmI>
> .
>

Garyfallidis commented:
Hi @rutgerfick @arokem and @demianw. I would love to see this rebased and ready to be merged. Can we get back to it please? Many people would love to start using this great method and its updates. Thanks in advance.

rutgerfick commented:
Hey Elef!!

Great to hear from you again!!
I want to let you know that finishing this contribution is a high priority
for me.
I'm currently preparing my thesis manuscript will finish in the next couple
of weeks.
Once this is done I will make the final corrections to the code and the
example file, as per your instructions, so it can finally be merged :)

Without dipy I definitely could not have accomplished as much as I have
during my PhD. My sincere thanks for your vision!

I'll get back to you soon,

Rutger


On 23 November 2016 at 01:25, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> Hi @rutgerfick <https://github.com/rutgerfick> @arokem
> <https://github.com/arokem> and @demianw <https://github.com/demianw>. I
> would love to see this rebased and ready to be merged. Can we get back to
> it please? Many people would love to start using this great method and its
> updates. Thanks in advance.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/740#issuecomment-262406278>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI-37Lfb2r6_qF7krbWM9avjeu_mpLQrks5rA4gJgaJpZM4GRWmI>
> .
>

arokem commented:
@Garyfallidis : what comments are still pending once this is rebased? 

I created a rebased version of this on #1153

arokem commented:
@Garyfallidis : does anything else need to happen apart from the rebase? I have a rebased version of this in #1153

arokem commented:
Superseded by #1153
